ui/ducks/uns.js

```diff
@@ -0,0 +1,217 @@
+          state.error = UNS_COMMON_ERROR;
+        } else {
+          state.resolution = address;
+          state.domainName = unsName;
+        }
+        if (isConfusing(unsName)) {
+          state.warning = UNS_CONFUSING_ERROR;
+        }
+      } else {
+        state.error = UNS_COMMON_ERROR;
+      }
+    },
+    enableUnsLookup: (state, action) => {
+      state.stage = 'INITIALIZED';
+      state.error = null;
+      state.resolution = null;
+      state.warning = null;
+      state.network = action.payload;
+    },
+    resetUnsResolution: (state) => {
+      state.resolution = null;
+      state.warning = null;
+      state.error = null;
+    },
+  },
+  extraReducers: (builder) => {
+    builder.addCase(CHAIN_CHANGED, (state, action) => {
+      if (action.payload !== state.currentChainId) {
+        state.stage = 'UNINITIALIZED';
+      }
+    });
+  },
+});
+
+const { reducer, actions } = slice;
+export default reducer;
+
+const {
+  unsLookup,
+  enableUnsLookup,
+  resetUnsResolution,
+} = actions;
+export { resetUnsResolution, unsLookup }
+
+export function initializeUnsSlice() {
+  return (dispatch, getState) => {
+    const state = getState();
+    const chainId = getCurrentChainId(state);
+    const network = CHAIN_ID_TO_NETWORK_ID_MAP[chainId];
+    dispatch(enableUnsLookup(network));
+  };
+}
+
+export function prepareResolutionCall(unsName) {
+  return async (dispatch, getState) => {
+    let state = getState();
+    let result;
+    if (state[name].stage === 'UNINITIALIZED') {
+      await dispatch(initializeUnsSlice());
+    }
+    if (state.send.draftTransactions[state.send.currentTransactionUUID].asset.type === 'NATIVE') {
+      result = await determineChainType(state.metamask.nativeCurrency);
+    } else {
+      result = 'MULTI_CHAIN';
+    }
+    if (result === 'SINGLE_CHAIN') {
+      let object = await resolveUNS(unsName, state.metamask.nativeCurrency);
+      await dispatch(
+        unsLookup({
+          unsName: object.unsName,
+          address: object.address,
+          error: object.error,
+        }),
+      );
+    } else if (result === 'MULTI_CHAIN') {
+      let object = {};
+      if (state.send.draftTransactions[state.send.currentTransactionUUID].asset.type === 'NATIVE') {
+        object = await resolveMultiChainUNS(unsName, state.metamask.nativeCurrency, state.metamask.nativeCurrency);
+      } else {
+        object = await resolveMultiChainUNS(unsName, state.send.draftTransactions[state.send.currentTransactionUUID].asset.details.symbol, state.send.draftTransactions[state.send.currentTransactionUUID].asset.details.standard);
+      }
+      await dispatch(
+        unsLookup({
+          unsName: object.unsName,
+          address: object.address,
+          error: object.error,
+        }),
+      );
+    }
+  }
+}
+
+export async function swapToken(unsName, asset) {
+  let object = {}
+  if (typeof (asset) === 'string' || typeof (asset) === 'object') {
```

Comment:
Why or `typeof (asset) === 'object'`? If it's an object then wouldn't it fail when calling `determineChainType()` and `resolveUNS()` as they're expecting `string` type, right?

Reply:
it's a little weird. I've tested and we can definitely remove the IF statement however, the way MetaMask handles different currencies asset can be both a String (for native currencies like ETH) and an Object (for multichain tokens like USDT).

Reply:
Ah I see. Then I would add this if before calling `determineChainType()`, and the else will straight go to the `object = await resolveMultiChainUNS(unsName, asset.details.symbol, asset.details.standard);` statement